data/tutorials/language/3ds_07_monads.md

```diff
@@ -0,0 +1,822 @@
+---
+id: monads
+title: Monads
+description: >
+  Monads, adapted from OCaml Programming: Correct + Efficient + Beautiful
+category: "Data Structures"
+external_tutorial:
+  tag: "CS3110"
+  banner:
+    image: "tutorials/cs3110_banner.png"
+    url: https://cs3110.github.io/textbook/cover.html
+    alt: "Ocaml Programming: Correct + Efficient + Beautiful"
+  contribute_link:
+    url: https://github.com/cs3110/textbook/blob/main/src/chapters/ds/monads.md
+    description: "You are encouraged to contribute to the original sources of this page at the CS3110 GitHub repository."
+---
+
+This is an excerpt of the page ["Monads"](https://cs3110.github.io/textbook/chapters/ds/monads.html) from the book [OCaml Programming: Correct + Efficient + Beautiful](https://cs3110.github.io/textbook/cover.html), reproduced here with permission.
+
+# Monads
+
+A *monad* is more of a design pattern than a data structure. That is, there are
+many data structures that, if you look at them in the right way, turn out to be
+monads.
+
+The name "monad" comes from the mathematical field of *category theory*, which
+studies abstractions of mathematical structures. If you ever take a PhD level
+class on programming language theory, you will likely encounter that idea in
+more detail. Here, though, we will omit most of the mathematical theory and
+concentrate on code.
+
+Monads became popular in the programming world through their use in Haskell, a
+functional programming language that is even more pure than OCaml&mdash;that is,
+Haskell avoids side effects and imperative features even more than OCaml. But no
+practical language can do this without side effects. After all, printing to the
+screen is a side effect. So Haskell set out to control the use of side effects
+through the monad design pattern. Since then, monads have become recognised as
+useful in other functional programming languages, and they are even starting to
+appear in imperative languages.
```

Comment:
```suggestion
Monads became popular in the programming world through their use in Haskell, a nother 
functional programming language. Haskell avoids side effects and imperative features even more than OCaml. But no
practical language can do this without side effects. After all, printing to the
screen is a side effect. So Haskell attempts to control the use of side effects
through the monad design pattern. Since then, monads have become recognised as
useful in other functional programming languages, and they are even starting to
appear in imperative languages.
```